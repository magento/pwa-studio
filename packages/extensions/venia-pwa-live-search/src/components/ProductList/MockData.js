/*
Copyright 2024 Adobe
All Rights Reserved.

NOTICE: Adobe permits you to use, modify, and distribute this file in
accordance with the terms of the Adobe license agreement accompanying
it.
*/

const SimpleProduct = {
    product: {
        sku: '24-WG088',
        name: 'Sprite Foam Roller',
        canonical_url:
            'http://master-7rqtwti-grxawiljl6f4y.us-4.magentosite.cloud/sprite-foam-roller.html'
    },
    productView: {
        __typename: 'SimpleProductView',
        sku: '24-WG088',
        name: 'Sprite Foam Roller',
        url:
            'http://master-7rqtwti-grxawiljl6f4y.us-4.magentosite.cloud/sprite-foam-roller.html',
        urlKey: 'sprite-foam-roller',
        images: [
            {
                label: 'Image',
                url:
                    'http://master-7rqtwti-grxawiljl6f4y.us-4.magentosite.cloud/media/catalog/product/l/u/luma-foam-roller.jpg'
            }
        ],
        price: {
            final: {
                amount: {
                    value: 19.0,
                    currency: 'USD'
                }
            },
            regular: {
                amount: {
                    value: 19.0,
                    currency: 'USD'
                }
            }
        }
    },
    highlights: [
        {
            attribute: 'name',
            value: '<em>Sprite</em> Foam Roller',
            matched_words: []
        }
    ]
};

const ComplexProduct = {
    product: {
        sku: 'MSH06',
        name: 'Lono Yoga Short',
        canonical_url:
            'http://master-7rqtwti-grxawiljl6f4y.us-4.magentosite.cloud/lono-yoga-short.html'
    },
    productView: {
        __typename: 'ComplexProductView',
        sku: 'MSH06',
        name: 'Lono Yoga Short',
        url:
            'http://master-7rqtwti-grxawiljl6f4y.us-4.magentosite.cloud/lono-yoga-short.html',
        urlKey: 'lono-yoga-short',
        images: [
            {
                label: '',
                url:
                    'http://master-7rqtwti-grxawiljl6f4y.us-4.magentosite.cloud/media/catalog/product/m/s/msh06-gray_main_2.jpg'
            },
            {
                label: '',
                url:
                    'http://master-7rqtwti-grxawiljl6f4y.us-4.magentosite.cloud/media/catalog/product/m/s/msh06-gray_alt1_2.jpg'
            },
            {
                label: '',
                url:
                    'http://master-7rqtwti-grxawiljl6f4y.us-4.magentosite.cloud/media/catalog/product/m/s/msh06-gray_back_2.jpg'
            }
        ],
        priceRange: {
            maximum: {
                final: {
                    amount: {
                        value: 32.0,
                        currency: 'USD'
                    }
                },
                regular: {
                    amount: {
                        value: 32.0,
                        currency: 'USD'
                    }
                }
            },
            minimum: {
                final: {
                    amount: {
                        value: 32.0,
                        currency: 'USD'
                    }
                },
                regular: {
                    amount: {
                        value: 32.0,
                        currency: 'USD'
                    }
                }
            }
        },
        options: [
            {
                id: 'size',
                title: 'Size',
                values: [
                    {
                        title: '32',
                        id: 'Y29uZmlndXJhYmxlLzE4Ni8xODQ=',
                        type: 'TEXT',
                        value: '32'
                    },
                    {
                        title: '33',
                        id: 'Y29uZmlndXJhYmxlLzE4Ni8xODU=',
                        type: 'TEXT',
                        value: '33'
                    },
                    {
                        title: '34',
                        id: 'Y29uZmlndXJhYmxlLzE4Ni8xODY=',
                        type: 'TEXT',
                        value: '34'
                    },
                    {
                        title: '36',
                        id: 'Y29uZmlndXJhYmxlLzE4Ni8xODc=',
                        type: 'TEXT',
                        value: '36'
                    }
                ]
            },
            {
                id: 'color',
                title: 'Color',
                values: [
                    {
                        title: 'Blue',
                        id: 'Y29uZmlndXJhYmxlLzkzLzU5',
                        type: 'COLOR_HEX',
                        value: '#1857f7'
                    },
                    {
                        title: 'Red',
                        id: 'Y29uZmlndXJhYmxlLzkzLzY3',
                        type: 'COLOR_HEX',
                        value: '#ff0000'
                    },
                    {
                        title: 'Gray',
                        id: 'Y29uZmlndXJhYmxlLzkzLzYx',
                        type: 'COLOR_HEX',
                        value: '#8f8f8f'
                    }
                ]
            }
        ]
    },
    highlights: [
        {
            attribute: 'name',
            value: 'Lono <em>Yoga</em> Short',
            matched_words: []
        }
    ]
};

export const products = [
    SimpleProduct,
    SimpleProduct,
    SimpleProduct,
    SimpleProduct,
    SimpleProduct,
    SimpleProduct,
    ComplexProduct
];
