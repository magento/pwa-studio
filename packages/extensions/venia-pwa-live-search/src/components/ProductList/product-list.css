/* https://cssguidelin.es/#bem-like-naming */
.sfsdk-product-list {
}

/* Extra small devices (phones, 600px and down) */
/* @media only screen and (max-width: 600px) { } */

/* Small devices (portrait tablets and large phones, 600px and up) */
/* @media only screen and (min-width: 600px) { } */

/* Medium devices (landscape tablets, 768px and up) */
/* @media only screen and (min-width: 768px) { } */

/* Large devices (laptops/desktops, 992px and up) */
/* @media only screen and (min-width: 992px) { } */

/* Extra large devices (large laptops and desktops, 1200px and up) */
/* @media only screen and (min-width: 1200px) { } */
