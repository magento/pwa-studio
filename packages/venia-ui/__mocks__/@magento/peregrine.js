import React from 'react';
import { BrowserPersistence } from './util/simplePersistence';

const peregrine = jest.requireActual('../../../peregrine/lib');

// re-exports
const RestApi = {
    Magento2: {
        request: jest.fn()
    }
};

const Util = { BrowserPersistence };

// hooks
const useEventListener = jest.fn(peregrine.useEventListener);
const useDropdown = jest.fn(peregrine.useDropdown);
const useSearchParam = jest.fn(peregrine.useSearchParam);
const useWindowSize = jest.fn(peregrine.useWindowSize);
const useToasts = jest.fn(peregrine.useToasts);
const useTypePolicies = jest.fn(peregrine.useTypePolicies);
// components

/**
 * the Price component from @magento/peregrine
 * has browser-specific functionality and cannot
 * currently by rendered in the test environment
 */
const Price = jest.fn().mockReturnValue(<div />);

module.exports = {
    ...peregrine,
    Price,
    RestApi,
    Util,
    useEventListener,
    useDropdown,
    useSearchParam,
    useWindowSize,
    useToasts,
    useTypePolicies
};
