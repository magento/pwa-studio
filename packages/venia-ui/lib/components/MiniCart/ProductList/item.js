import React, { useMemo } from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { string, number, shape, func, arrayOf, oneOf } from 'prop-types';
import { Trash2 as DeleteIcon } from 'react-feather';
import { Link } from 'react-router-dom';

import Price from '@magento/venia-ui/lib/components/Price';
import { useItem } from '@magento/peregrine/lib/talons/MiniCart/useItem';
import resourceUrl from '@magento/peregrine/lib/util/makeUrl';

import ProductOptions from '../../LegacyMiniCart/productOptions';
import Image from '../../Image';
import Icon from '../../Icon';
import { useStyle } from '../../../classify';
import configuredVariant from '@magento/peregrine/lib/util/configuredVariant';

import defaultClasses from './item.css';

const Item = props => {
    const {
        classes: propClasses,
        product,
        id,
        quantity,
        configurable_options,
        handleRemoveItem,
        prices,
        closeMiniCart,
        configurableThumbnailSource
    } = props;

    const { formatMessage } = useIntl();
    const classes = useStyle(defaultClasses, propClasses);
    const itemLink = useMemo(
        () => resourceUrl(`/${product.url_key}${product.url_suffix || ''}`),
        [product.url_key, product.url_suffix]
    );
    const stockStatusText =
        product.stock_status === 'OUT_OF_STOCK'
            ? formatMessage({
                  id: 'productList.outOfStock',
                  defaultMessage: 'Out-of-stock'
              })
            : '';

    const { isDeleting, removeItem } = useItem({
        id,
        handleRemoveItem
    });

    const rootClass = isDeleting ? classes.root_disabled : classes.root;
    const configured_variant = configuredVariant(configurable_options, product);

    return (
        <div className={rootClass}>
            <Link
                className={classes.thumbnailContainer}
                to={itemLink}
                onClick={closeMiniCart}
            >
                <Image
                    alt={product.name}
                    classes={{
                        root: classes.thumbnail
                    }}
                    width={100}
                    resource={
                        configurableThumbnailSource === 'itself' &&
                        configured_variant
                            ? configured_variant.thumbnail.url
                            : product.thumbnail.url
                    }
                />
            </Link>
            <Link
                className={classes.name}
                to={itemLink}
                onClick={closeMiniCart}
            >
                {product.name}
            </Link>
            <ProductOptions
                options={configurable_options}
                classes={{
                    options: classes.options
                }}
            />
            <span className={classes.quantity}>
                <FormattedMessage
                    id={'productList.quantity'}
                    defaultMessage={'Qty :'}
                    values={{ quantity }}
                />
            </span>
            <span className={classes.price}>
                <Price
                    currencyCode={prices.price.currency}
                    value={prices.price.value}
                />
                <FormattedMessage
                    id={'productList.each'}
                    defaultMessage={' ea.'}
                />
            </span>
            <span className={classes.stockStatus}>{stockStatusText}</span>
            <button
                onClick={removeItem}
                type="button"
                className={classes.deleteButton}
                disabled={isDeleting}
            >
                <Icon
                    size={16}
                    src={DeleteIcon}
                    classes={{
                        icon: classes.editIcon
                    }}
                />
            </button>
        </div>
    );
};

export default Item;

Item.propTypes = {
    classes: shape({
        root: string,
        thumbnail: string,
        name: string,
        options: string,
        quantity: string,
        price: string,
        editButton: string,
        editIcon: string
    }),
    product: shape({
        name: string,
        thumbnail: shape({
            url: string
        })
    }),
    id: string,
    quantity: number,
    configurable_options: arrayOf(
        shape({
            id: number,
            option_label: string,
            value_id: number,
            value_label: string
        })
    ),
    handleRemoveItem: func,
    prices: shape({
        price: shape({
            value: number,
            currency: string
        })
    }),
    configured_variant: shape({
        thumbnail: shape({
            url: string
        })
    }),
    configurableThumbnailSource: oneOf(['parent', 'itself'])
};
