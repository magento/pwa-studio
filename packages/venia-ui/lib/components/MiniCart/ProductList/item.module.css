.root {
    composes: gap-x-xs from global;
    composes: gap-y-2xs from global;
    composes: grid from global;
    composes: opacity-100 from global;
    composes: text-sm from global;
    grid-template-columns: auto 1fr auto;
    grid-template-rows: auto auto auto auto auto 1fr;
    transition-delay: 64ms;
    transition-duration: 384ms;
    transition-property: opacity;
    transition-timing-function: var(--venia-global-anim-standard);
}

.root_disabled {
    composes: root;

    composes: opacity-50 from global;
}

.thumbnailContainer {
    grid-column: 1 / span 1;
    grid-row: 1 / span 6;
}

.thumbnail {
    composes: border from global;
    composes: border-solid from global;
    composes: border-subtle from global;
}

.name {
    composes: font-semibold from global;
    composes: mb-2xs from global;
    grid-column: 2 / span 1;
    grid-row: 1 / span 1;
}

.options {
    composes: gap-2xs from global;
    composes: grid from global;
    grid-column: 2 / span 1;
    grid-row: 2 / span 2;
}

.quantity {
    grid-column: 2 / span 1;
    grid-row: 4 / span 1;
}

.price {
    grid-column: 2 / span 1;
    grid-row: 5 / span 1;
}

.stockStatus {
    composes: font-semibold from global;
    composes: text-error from global;
    grid-column: 2;
    grid-row: 6;
}

.deleteButton {
    grid-column: 3 / span 1;
    grid-row: 1 / span 1;
}
