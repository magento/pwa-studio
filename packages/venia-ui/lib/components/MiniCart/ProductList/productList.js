import React, { useMemo } from 'react';
import { string, func, arrayOf, shape, number, oneOf } from 'prop-types';

import Item from './item';
import { useStyle } from '../../../classify';

import defaultClasses from './productList.css';

const ProductList = props => {
    const {
        items,
        handleRemoveItem,
        classes: propClasses,
        closeMiniCart,
        configurableThumbnailSource
    } = props;
    const classes = useStyle(defaultClasses, propClasses);

    const cartItems = useMemo(() => {
        if (items) {
            return items.map(item => (
                <Item
                    key={item.id}
                    {...item}
                    closeMiniCart={closeMiniCart}
                    handleRemoveItem={handleRemoveItem}
                    configurableThumbnailSource={configurableThumbnailSource}
                />
            ));
        }
    }, [items, handleRemoveItem, closeMiniCart, configurableThumbnailSource]);

    return <div className={classes.root}>{cartItems}</div>;
};

export default ProductList;

ProductList.propTypes = {
    classes: shape({ root: string }),
    items: arrayOf(
        shape({
            product: shape({
                name: string,
                thumbnail: shape({
                    url: string
                })
            }),
            id: string,
            quantity: number,
            configurable_options: arrayOf(
                shape({
                    label: string,
                    value: string
                })
            ),
            prices: shape({
                price: shape({
                    value: number,
                    currency: string
                })
            }),
            configured_variant: shape({
                thumbnail: shape({
                    url: string
                })
            })
        })
    ),
    configurableThumbnailSource: oneOf(['parent', 'itself']),
    handleRemoveItem: func
};
