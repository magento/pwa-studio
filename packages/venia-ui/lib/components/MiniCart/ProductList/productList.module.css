.root {
    display: grid;
    gap: 1.5rem;
}
