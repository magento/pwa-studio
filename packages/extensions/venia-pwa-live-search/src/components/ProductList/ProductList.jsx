/*
Copyright 2024 Adobe
All Rights Reserved.

NOTICE: Adobe permits you to use, modify, and distribute this file in
accordance with the terms of the Adobe license agreement accompanying
it.
*/

import React, { useEffect, useState, useMemo } from 'react';
import './product-list.css';

import { Alert } from '../../components/Alert';
import { useProducts, useStore } from '../../context';
import { ProductItem } from '../ProductItem';
import { classNames } from '../../utils/dom';
import { sanitizeProducts } from '../../utils/modifyResults';
import { useResultsModifier } from '../../context/resultsModifierContext';


const ProductList = ({ products, numberOfColumns, showFilters }) => {
  const productsCtx = useProducts();
  const {
    currencySymbol,
    currencyRate,
    setRoute,
    refineProduct,
    refreshCart,
    addToCart,
  } = productsCtx;

  const [cartUpdated, setCartUpdated] = useState(false);
  const [itemAdded, setItemAdded] = useState('');
  const { viewType } = useProducts();
  const [error, setError] = useState(false);

  const {
    config: { listview },
  } = useStore();

  // const cleanUrl = (url, cleanWithProtocol = false) => {
  //   if (!url) return url;
  //   try {
  //     if (cleanWithProtocol) return url.replace('http://local.magentocomposeree.com:8082', '');
  //     return url.replace('//local.magentocomposeree.com:8082', '');
  //   } catch {
  //     return url;
  //   }
  // };

  const { baseUrl, baseUrlWithoutProtocol } = useResultsModifier();
  //console.log("PROductList baseURL = ",baseUrl);
  //console.log("PROductList baseUrlWithoutProtocol = ",baseUrlWithoutProtocol);

  const sanitizedProducts = useMemo(() => sanitizeProducts(products, baseUrl, baseUrlWithoutProtocol), [products]);  
//   const sanitizedProducts = useMemo(() => {
//   if (!products || products.length === 0) return [];

//   return products.map(item => {
//     const prod = item.product;
//     const prodView = item.productView;

//     if (!prod) return item;

//     const sanitizeImage = (img, cleanWithProtocol = false) => {
//       if (Array.isArray(img)) {
//         return img.map(i =>
//           i && i.url ? { ...i, url: cleanUrl(i.url, cleanWithProtocol) } : i
//         );
//       }

//       return img && img.url ? { ...img, url: cleanUrl(img.url, cleanWithProtocol) } : img;
//     };

//     return {
//       ...item,
//       product: {
//         ...prod,
//         canonical_url: cleanUrl(prod.canonical_url),
//         image: sanitizeImage(prod.image),
//         small_image: sanitizeImage(prod.small_image),
//         thumbnail: sanitizeImage(prod.thumbnail)
//       },
//       productView: {
//         ...prodView,
//         images: sanitizeImage(prodView?.images, true),
//         url: cleanUrl(prodView?.url, true)
//       }
//     };
//   });
// }, [products]);

  // console.log("sanitizedProducts = ",sanitizedProducts);
  // console.log("products = ",JSON.stringify(products));
  // console.log("ProductList.jsx refineProduct = ",refineProduct);

  const className = showFilters
    ? 'ds-sdk-product-list bg-body max-w-full pl-3 pb-2xl sm:pb-24'
    : 'ds-sdk-product-list bg-body w-full mx-auto pb-2xl sm:pb-24';

  useEffect(() => {
    if (refreshCart) refreshCart();
  }, [itemAdded]);

  return (
    <div
      className={classNames(
        'ds-sdk-product-list bg-body pb-2xl sm:pb-24',
        className
      )}
    >
      {cartUpdated && (
        <div className="mt-8">
          <Alert
            title={`You added ${itemAdded} to your shopping cart.`}
            type="success"
            description=""
            onClick={() => setCartUpdated(false)}
          />
        </div>
      )}
      {error && (
        <div className="mt-8">
          <Alert
            title="Something went wrong trying to add an item to your cart."
            type="error"
            description=""
            onClick={() => setError(false)}
          />
        </div>
      )}

      {listview && viewType === 'listview' ? (
        <div className="w-full">
          <div className="ds-sdk-product-list__list-view-default mt-md grid grid-cols-none pt-[15px] w-full gap-[10px]">
            {sanitizedProducts?.map((product) => (
              //console.log("listview = ", product)
              <ProductItem
                key={product?.productView?.id}
                item={product}
                currencySymbol={currencySymbol}
                currencyRate={currencyRate}
                setRoute={setRoute}
                refineProduct={refineProduct}
                setCartUpdated={setCartUpdated}
                setItemAdded={setItemAdded}
                addToCart={addToCart}
                setError={setError}
              />
            ))}
          </div>
        </div>
      ) : (
        <div
          style={{
            gridTemplateColumns: `repeat(${numberOfColumns}, minmax(0, 1fr))`,
          }}
          className="ds-sdk-product-list__grid mt-md grid gap-y-8 gap-x-2xl xl:gap-x-8"
        >
          {sanitizedProducts?.map((product) => (
            //console.log("non list view = ", product)
            <ProductItem
              key={product?.productView?.id}
              item={product}
              currencySymbol={currencySymbol}
              currencyRate={currencyRate}
              setRoute={setRoute}
              refineProduct={refineProduct}
              setCartUpdated={setCartUpdated}
              setItemAdded={setItemAdded}
              addToCart={addToCart}
              setError={setError}
            />
          ))}
        </div>
      )}
    </div>
  );
};

export default ProductList;
