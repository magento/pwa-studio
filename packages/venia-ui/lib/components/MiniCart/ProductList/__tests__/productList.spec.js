import React from 'react';
import { createTestInstance } from '@magento/peregrine';

import ProductList from '../productList';

jest.mock('../../../../classify');
jest.mock('../item', () => 'Item');
jest.mock('react-router-dom', () => ({
    Link: ({ children, ...rest }) => <div {...rest}>{children}</div>
}));
jest.mock('@magento/peregrine/lib/util/makeUrl');

const props = {
    closeMiniCart: jest.fn().mockName('closeMiniCart'),
    handleRemoveItem: jest.fn().mockName('handleRemoveItem'),
    items: [
        {
            id: '1',
            product: {
                name: 'Simple Product'
            }
        }
    ]
};

test('Should render properly', () => {
    const tree = createTestInstance(<ProductList {...props} />);

    expect(tree.toJSON()).toMatchSnapshot();
});
