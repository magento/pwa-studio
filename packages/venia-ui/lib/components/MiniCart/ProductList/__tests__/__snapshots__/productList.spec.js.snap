// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render properly 1`] = `
<div
  className="root"
>
  <Item
    closeMiniCart={[MockFunction closeMiniCart]}
    handleRemoveItem={[MockFunction handleRemoveItem]}
    product={
      Object {
        "name": "Simple Product",
      }
    }
    uid="1"
  />
</div>
`;
